airbyte-cdk/python/airbyte_cdk/sources/streams/http/error_handlers/http_status_error_handler.py

```diff
@@ -0,0 +1,108 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+import logging
+from typing import Any, Mapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.models import FailureType
+
+from .error_handler import ErrorHandler
+from .response_action import ResponseAction
+
+
+class HttpStatusErrorHandler(ErrorHandler):
+
+    _DEFAULT_ERROR_MAPPING: Mapping[Union[int, str, Exception], Mapping[str, Any]] = {
+        400: {
+            "action": ResponseAction.FAIL,
+            "failure_type": FailureType.config_error,
+            "error_message": "Bad request. Please check your request parameters.",
+        },
+        401: {
+            "action": ResponseAction.FAIL,
+            "failure_type": FailureType.config_error,
+            "error_message": "Unauthorized. Please ensure you are authenticated correctly.",
+        },
+        403: {
+            "action": ResponseAction.FAIL,
+            "failure_type": FailureType.config_error,
+            "error_message": "Forbidden. You don't have permission to access this resource.",
+        },
+        404: {
+            "action": ResponseAction.FAIL,
+            "failure_type": FailureType.system_error,
+            "error_message": "Not found. The requested resource was not found on the server.",
+        },
+        408: {
+            "action": ResponseAction.FAIL,
+            "failure_type": FailureType.transient_error,
+            "error_message": "Request timeout. Please try again later.",
+        },
+        429: {
+            "action": ResponseAction.FAIL,
```

Comment:
Working on Salesforce right now and checking this, I'm not sure how practical having the same retry logic for any errors makes sense. Let's say I have a rate limiting window of 60 minutes, that seems fine to retry for 60 minutes on those errors but would we retry for 60 minutes on other transient errors (maybe?) or on unknown errors (like the one mentioned [here](https://github.com/airbytehq/airbyte/pull/37576/files#r1591583211))? 

Like, I don't see any problem on the server side to get retried on client errors like this but it seems like it could delay sync failure for no apparent reason. I fear for cases where the rate limiting will have a very long window as we will need to align all our retries with this very long rolling window...

Reply:
longer discussion thread on slack: https://airbytehq-team.slack.com/archives/C02TL38U5L7/p1715188986172059

I think we agree that this is a valid feature, but not strictly needed for the initial implementation. And this is considered additive because we've only ever supported the configurable backoff at the stream level before now.

I wrote the initial ticket here, but still needs to be scoped a bit more https://github.com/airbytehq/airbyte-internal-issues/issues/7646